### PR TITLE
[R-package] fixed best_iter and best_score when training data is passed (fixes #2295, #2525)

### LINKED ISSUE #2295: Error in best score and best iter, wrong value returned
## Environment info

Ubuntu 18.04.2 LTS (GNU/Linux 4.18.0-1024-azure x86_64)

R 3.6.0 lightgbm version 2.2.4

LightGBM version or commit hash:

## Error message

There is no error message but the values returned by best_iter and best_score are wrong when the training dataset is part of the valids.

This appears to be related to the May 27 2019 commit to fix best_iter and best_score and those new lines:

`if (record && is.na(env$best_score)) {
    if (env$eval_list[[1]]$higher_better[1] == TRUE) {
      cv_booster$best_iter <- unname(which.max(unlist(cv_booster$record_evals[[2]][[1]][[1]])))
      cv_booster$best_score <- cv_booster$record_evals[[2]][[1]][[1]][[cv_booster$best_iter]]
    } else {
      cv_booster$best_iter <- unname(which.min(unlist(cv_booster$record_evals[[2]][[1]][[1]])))
      cv_booster$best_score <- cv_booster$record_evals[[2]][[1]][[1]][[cv_booster$best_iter]]
    }
  }`

`  # When early stopping is not activated, we compute the best iteration / score ourselves by selecting the first metric and the first dataset
  if (record && length(valids) > 0 && is.na(env$best_score)) {
    if (env$eval_list[[1]]$higher_better[1] == TRUE) {
      booster$best_iter <- unname(which.max(unlist(booster$record_evals[[2]][[1]][[1]])))
      booster$best_score <- booster$record_evals[[2]][[1]][[1]][[booster$best_iter]]
    } else {
      booster$best_iter <- unname(which.min(unlist(booster$record_evals[[2]][[1]][[1]])))
      booster$best_score <- booster$record_evals[[2]][[1]][[1]][[booster$best_iter]]
    }
  }
`

Lightgbm itself behaves correctly as the early stopping parameter is applied correctly as per the documentation on lightgbm. 

"Activates early stopping. The model will train until the validation score stops improving. Validation score needs to improve at least every early_stopping_rounds round(s) to continue training. Requires at least one validation data and one metric. If there’s more than one, will check all of them. But the training data is ignored anyway. To check only the first metric set first_metric_only to True."

The training data is ignored for early_stopping, so best_score and best_iter should not apply to the train data.

If we look at xgboost documentation for a comparison we have this:

"early_stopping_rounds (int) – Activates early stopping. Validation metric needs to improve at least once in every early_stopping_rounds round(s) to continue training. Requires at least one item in evals. The method returns the model from the last iteration (not the best one). If there’s more than one item in evals, the last entry will be used for early stopping."

While the early stop implementation in lightgbm uses all valids except the train data unlike xgboost, modifying the code to use the last valids dataset instead of the first one for best_iter and best_score  would likely prevent best_iter and best_score being associated with the train data.

## Reproducible examples

`#Starting from the lgb.train example

library(lightgbm)
data(agaricus.train, package = "lightgbm")
train <- agaricus.train
dtrain <- lgb.Dataset(train$data, label = train$label)
data(agaricus.test, package = "lightgbm")
test <- agaricus.test
dtest <- lgb.Dataset.create.valid(dtrain, test$data, label = test$label)
params <- list(objective = "regression", metric = "l2")
valids <- list(test = dtest)
model <- lgb.train(params,
                   dtrain,
                   100,
                   valids,
                   min_data = 1,
                   learning_rate = 1,
                   early_stopping_rounds = 10)

#Early stop works correctly and the value returned by best score and best iter is correct.

model$best_iter
model$best_score

#However, if we add the train as part of the valids to see how it improves
#comparatively to the test to try to see if the model is overfitting,
#the best_score and best_iter are based on the train, which makes no sense.

valids2<-list(train=dtrain,test=dtest)

model2<-lgb.train(params,
                  dtrain,
                  100,
                  valids2,
                  min_data = 1,
                  learning_rate = 1,
                  early_stopping_rounds = 10)

model2$best_iter
which.min(unlist(model2$record_evals$test$l2$eval))
which.min(unlist(model2$record_evals$train$l2$eval))

#This might be expected since the fix of May 27 2019 on used the first valid available when env$best_score and env$best_iter was NA
#However, if we try to move the test first, the train always ends up being the first parameter so the best_iter and best_score are always wrong

valids3<-list(test=dtest,train=dtrain)

model3<-lgb.train(params,
                  dtrain,
                  100,
                  valids3,
                  min_data = 1,
                  learning_rate = 1,
                  early_stopping_rounds = 10)

model3$best_iter
which.min(unlist(model3$record_evals$test$l2$eval))
which.min(unlist(model3$record_evals$train$l2$eval))`

### PR DESCRIPTION
**Pull request by jameslamb**
I started to pick up #2471 but found that I needed to make enough changes that I thought it made sense to just put them in a new PR and close that one. I also think this makes the review process cleaner, so I'm not technically updating a PR that belongs to @Laurae2 while also acting as a reviewer on my own code.

I tried to address #2295 in this PR. Basically, when early stopping is not used, the best score and best iteration should be determined using the first metric and first validation set given, **excluding the training data**. #2295 suggested that right now, if you pass the training data to `valids` it's possible that `best_iter` and  `best_score` will be based on the training data.

This PR proposes the following improvements:

* exclude training data from consideration for `best_iter` and `best_score` in `lgb.train()`
* replace positional references like `booster$record_evals[[2L]][[1L]][[1L]]` with the actual names used in those nested lists
* centralize the naming scheme for `record_evals` in package constants to reduce risk of silent errors caused by hard-coding 
* add tests on `best_score` and `best_iter` behavior for classification and regression, for metrics to maximize and those to minimize, and for cases where the training data is passed to `valids` with a name other than `"train"` (to be sure code detecting it with `identical()` works)

This PR does not attempt to fix https://github.com/microsoft/LightGBM/issues/2472, as noted in https://github.com/microsoft/LightGBM/pull/2471#issuecomment-536300877. That will be fixed in a later PR.

### CODE REVIEW

**Comment by jameslamb** [comment]
Gah our CI is  🔥🔥🔥 this week! 

From [Azure DevOps](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=5679&view=logs&j=a0acac97-df52-5651-dae8-9fb37a4728fc&t=d2c32b51-981b-5bcd-fe31-029dd8d8998c)

image

image

**Comment by jameslamb** [comment]
close re-open to re-trigger CI and check if the failures are transient

**Comment by jameslamb** [comment]
After reading #2525, I think this PR will fix that too

**Comment by jameslamb** [comment]
I've rebased this to update it with latest `master`. This PR is ready for review.

**Comment by jameslamb** [comment]
Thanks for the review @guolinke ! I'm glad to get this one merged, think it will improve the user experience
